src/cattributes.c

```diff
@@ -23,10 +23,10 @@
 
 #include "igraph_attributes.h"
 #include "igraph_memory.h"
-#include "config.h"
 #include "igraph_math.h"
 #include "igraph_interface.h"
 #include "igraph_random.h"
+#include "config.h"
```

Comment:
Be careful with moving the include of `config.h` around; it might be the case that some of the headers _below_ its original location but _above_ the current one depended on the presence of macros defined in `config.h`. If we decide to move `config.h` around, I'd probably move it _above_ all igraph-related includes.

Reply:
My original thought was that in that case the header itself should include `config.h`, thus if moving `config.h` to the end causes errors, then we found a bug that needs to be fixed. This is generally the reasoning behind moving system headers to the end: this practical reveals cases when a system header should have been included in a library header, but it wasn't.

But let me think for a moment because `config.h` is different.  I am going to be thinking aloud a but below. Please correct any mistaken assumptions.

`config.h` does not get installed (`make install`) so no public header should depend on it. Normally, any public header that may depend on thing in `config.h` is generated from an `.in` files (like `igraph_threading.h.in`).

I `grep`ped the `include` directory for `HAVE_` and found just one header: `igraph_vector.h` contains:

```
#ifdef HAVE_STDINT_H
    #include <stdint.h>
#else
    #if defined(HAVE_SYS_INT_TYPES_H) && HAVE_SYS_INT_TYPES_H
        #include <sys/int_types.h>    /* for Solaris */
    #endif
#endif
```

@ntamas Do you know if this is fine? None of these `HAVE_` macros will be defined when using igraph as a library. They would only be defined when compiling igraph itself. Is this intentional? Or is this a bug? It looks like a bug to me.

----

Now let's look at internal headers:

```
grep HAVE_ -R src --include '*.h*'

src/plfit/arithmetic_sse_float.h:#if     HAVE_XMMINTRIN_H
src/plfit/arithmetic_sse_float.h:#endif/*HAVE_XMMINTRIN_H*/
src/plfit/arithmetic_sse_double.h:#if     HAVE_EMMINTRIN_H
src/plfit/arithmetic_sse_double.h:#endif/*HAVE_EMMINTRIN_H*/
src/bliss/defs.hh:#if HAVE_GMP == 1
src/igraph_glpk_support.h:#ifdef HAVE_GLPK
src/igraph_math.h:#ifndef HAVE_LOG2
src/igraph_math.h:#ifndef HAVE_LOG1P
src/igraph_math.h:#ifndef HAVE_FABSL
src/igraph_math.h:#ifndef HAVE_FMIN
src/igraph_math.h:#ifndef HAVE_ROUND
src/igraph_hacks_internal.h:#ifndef HAVE_STRDUP
src/igraph_hacks_internal.h:#ifndef HAVE_STPCPY
src/igraph_hacks_internal.h:    #ifndef HAVE_STPCPY_SIGNATURE
src/igraph_arpack_internal.h:#ifdef HAVE_GFORTRAN
src/bignum.h:#ifdef HAVE_STDINT_H
src/bignum.h:    #ifdef HAVE_SYS_INT_TYPES_H
src/bignum.h:#ifndef HAVE_U64
src/bignum.h:    #define HAVE_U64 1
src/bignum.h:#if HAVE_U64
```

I verified that all of these internal headers have `#include "config.h"` at the beginning.

The one exception is `plfit`. However, those `HAVE_` macros are not in `config.h` and as far as I can tell will never be defined (??).  Is there anything that needs to be done here? As far as I can tell, those intrinsics are being used anyway.

Reply:
> `config.h` does not get installed (make install) so no public header should depend on it.

That's correct.

> Do you know if this is fine? None of these HAVE_ macros will be defined when using igraph as a library.

This is probably not fine; the question is whether there's anything in `igraph_vector.h` that needs `stdint.h` or `sys/int_types.h`. If there isn't, we can safely remove the whole block. If there is, we should probably find a way to work around it somehow. (But I think that this particular piece of code caters for an obscure use-case that I've found when I tried to get igraph compiled on an old Solaris server several years ago, and it might be the case that it's easier to say that we don't support old Solaris server builds any more).

> The one exception is `plfit`. However, those `HAVE_?MMINTRIN_H` macros are not in `config.h` and as far as I can tell will never be defined (??). 

These macros come from the upstream `plfit` distribution. The L-BFGS optimizer in `plfit` uses SSE2 optimized instructions if `emmintrin.h` is present. If it is not, it simply falls back to "normal" arithmetics. (See the `arithmetic_*.h` headers in `src/plfit`). So, this doesn't hurt, but we could improve things by also testing for `emmintrin.h` in the configure script and then allowing `src/plfit/lbfgs.c` to use SSE2 optimizations.